### ISSUE TITLE
Marking an existing Matrix field as translatable causes unexpected behavior with existing content

### ISSUE THREAD

**Issue by angrybrad** [craftcms/cms#2245]
### Steps to reproduce

1. Restore the attached database. [craft2.zip](https://github.com/craftcms/cms/files/1583297/craft2.zip)

username: admin, password: letmein

2. Notice there is an existing entry in the "News" section called "Existing Matrix Blocks translation test".  The entry has 3 Matrix blocks in both German and English.

3.  Go to "Einstellungen (Settings) -> Felder (Fields) -> Text", which is a Matrix field.

4.  Mark the entire field as translatable via "Das Feld kann übersetzt werden“ and save.

5.  Go back to the "Existing Matrix Blocks translation test“ entry.

6. Try to delete Matrix block from the German language version and save.

7.  The entry says it was saved successfully, but the block you just deleted is back.

8.  Additionally if you swap to the English version, an additional block has been duplicated into it.

### Additional info

- Craft version: 2.6.2300
- PHP version: 7.0
- Database driver & version: MySQL
- Plugins & versions: None

**Comment by cquinders** [comment]
I tested if this bug is also affecting Craft 3 RC3.

First I tried updating the database from [craft2.zip](https://github.com/craftcms/cms/files/1583297/craft2.zip) to Craft 3, and after that I recreated the test setting within a fresh installation of Craft 3 using the multi sites feature.

Both ways resulted in the same behaviour we are seeing in Craft 2.6.3000.

I can provide a database backup for testing this in Craft 3 if anybody is interested.

**Comment by cquinders** [comment]
@takobell @brandonkelly I just updated to Craft 2.6.3002 which includes the fix and tested the behaviour again. Unfortunately the problem still exists with my installation.

I performed the following steps:

1. Updated to Craft 2.6.3002 via the CP
2. Configured the Matrix field to be translateable
3. Opened an existing entry with blocks inside the matrix field
4. Added a new block in the foreign language and saved the entry
5. Switched to the main language. The added block from the foreign language is not there (as expected) but some blocks where duplicated in the main language
6. Additionally I tried deleting a block from the main language and saved the entry. Now the block was also missing from the foreign language (but it should have been preserved there)

**Comment by brandonkelly** [comment]
@cquinders Thanks for the report - fixed for good now!

**Comment by cquinders** [comment]
@brandonkelly I’m sorry, but this seems to be a really nasty bug. I tested again after installing Craft 2.6.3003 and still got issues with blocks being duplicated.

I tested multiple times and tried different variations, but in the end it comes down to this:

1. Updated to Craft 2.6.3003 via the CP
2. Configured the Matrix field to be translateable
2.1. [Without any effect: Opened the section settings in "Settings > Sections > ..." and hit "Save" in order to programmatically save each related entry again]
3. Opened an existing entry with blocks inside the matrix field
4. Saved the entry in any locale without changing anything
5. The content for this locale is unchanged (as expected)
6. Switched to a different locale of that entry
7. Every block in that locales matrix field was duplicated while saving the other locale

Do you think writing a script that re-saves the matrix field of every entry in every locale after configuring the matrix field to be translateable will be a possible work around here? I did not have the time to test this yet, but I might give it a try later on ...

**Comment by brandonkelly** [comment]
@cquinders I just re-tested and it’s working as expected for me.

> Every block in that locales matrix field was duplicated while saving the other locale

That is what is supposed to happen. When a Matrix field is translatable, each locale should get its own unique set of blocks. If you make an existing Matrix field translatable, then when you re-save an element that has it, it should first duplicate all of the existing blocks across all of the locales (so each locale has its own set of blocks, based on the original non-translatable blocks' content), and then it will update the _current_ locale with the new posted content.

**Comment by cquinders** [comment]
@brandonkelly Thanks for investigating again! What you described is what I was expecting as well.

Maybe I should have been more precise with what I mean with "duplicated". When I performed the steps described above (1. to 6.) I would see two copies of each previously contained matrix block inside the Matrix field. In one case the order of those blocks was sequential (same order as before, but two identical blocks instead of one) and in another case the additionally duplicated blocks were intermingled with the other blocks without any noticeable pattern.

**Comment by cquinders** [comment]
@brandonkelly In order to provide you with a little more context I tested further in Craft 2.6.3003 and here is what I found:

### Steps to reproduce with [craft2.zip](https://github.com/craftcms/cms/files/1583297/craft2.zip)

1. Restore the attached database [craft2.zip](https://github.com/craftcms/cms/files/1583297/craft2.zip) (username: admin, password: letmein)
2. Notice there is an existing entry in the "News" section called "Existing Matrix Blocks translation test". The entry has 3 Matrix blocks in both German and English. Also notice that textfields inside Matrix blocks are translatable and texts inside the English version are prepended with "English version: “.
3. Go to "Einstellungen (Settings) -> Felder (Fields) -> Text", which is a Matrix field.
4. Mark the entire field as translatable via "Das Feld kann übersetzt werden“ and save.
5. Go back to the "Existing Matrix Blocks translation test“ entry.
6. Save the entry in German without changing anything
7. The content for German is unchanged (as expected)
8. Switch to the English version
9. **Texts inside Matrix blocks are now missing the "English version: " part. This means they were overwritten with the German version.**

The issue with my production site (as described lastly) is a little bit different, but I think it fits the bigger picture:

### Steps with my production site

1. Configured the Matrix field to be translateable
2. Opened an existing entry with blocks inside the matrix field
3. Saved the entry in the main locale without changing anything
4. The content for this locale is unchanged (as expected)
5. Switched to a different locale of that entry
6. **Blocks from the main locale [de] where added to this locales [en] matrix field before the previously inherited block (e.g. A[de] - A[en] - B[de] - B[en] - ...).** It seems to make no difference if the fields inside those blocks were translated or still holding inherited content from the main locale.

### In summary

I suppose there is an issue with duplicating/branching existing matrix blocks into different locales after the matrix field was configured to be translatable. Either blocks in secondary locales are overwritten with contents from the main locale, or blocks from the main locale are added to secondary locales matrix field alongside existing blocks.

**Comment by angrybrad** [comment]
@cquinders this might sounds weird... but locally, can you disable devMode (if it is enabled) and see if it works as you'd expect?

**Comment by angrybrad** [comment]
@cquinders regardless, hopefully finally resolved in https://github.com/craftcms/cms/commit/bd764cdfde6388c0a74927471e50da54caf5ff81 now. Let us know if you still have issues after updating.

**Comment by cquinders** [comment]
@angrybrad I tried disabling devMode (which was enabled indeed) and it did work. Also after updating to 2.6.3006 the issue seems to be resolved.

Thanks for fixing this one!

**Comment by brandonkelly** [comment]
music to my ears :)

**Comment by angrybrad** [comment]
Third time is a charm!
